modules/charging.c

```diff
@@ -963,6 +971,214 @@ mch_dbus_get_charging_mode_cb(DBusMessage *const req)
     return TRUE;
 }
 
+/** Send charging_limit_enable D-Bus signal / method call reply
+ *
+ * @param req  method call message to reply, or NULL to send signal
+ */
+static void
+mch_dbus_send_charging_limit_enable(DBusMessage *const req)
+{
+    static int last = 0;
+
+    DBusMessage *msg = NULL;
+
+    dbus_int32_t value = mch_limit_enable;
+
+    if( req ) {
+        msg = dbus_new_method_reply(req);
+    }
+    else if( last == value ) {
+        goto EXIT;
+    }
+    else {
+        last = value;
+        msg = dbus_new_signal(MCE_SIGNAL_PATH,
+                              MCE_SIGNAL_IF,
+                              MCE_CHARGING_LIMIT_ENABLE_SIG);
+    }
+
+    if( !dbus_message_append_args(msg,
+                                  DBUS_TYPE_INT32, &value,
+                                  DBUS_TYPE_INVALID) )
+        goto EXIT;
+
+    mce_log(LL_DEBUG, "%s: %s = %d",
+            req ? "reply" : "broadcast",
+            "charging_limit_enable", value);
+
+    dbus_send_message(msg), msg = 0;
+
+EXIT:
+
+    if( msg )
+        dbus_message_unref(msg);
+}
+
+/** Callback for handling charging_limit_enable change request call
+ *
+ * @param req  Method call message to handle
+ *
+ * @return TRUE
+ */
+static gboolean
+mch_dbus_set_charging_limit_enable_cb(DBusMessage *const req)
+{
+    const char    *sender = dbus_message_get_sender(req);
+    DBusError      err    = DBUS_ERROR_INIT;
+    dbus_int32_t   limit  = 0;
+    DBusMessage   *rsp    = 0;
+
+    mce_log(LL_DEVEL, "enable-charging limit change request from %s",
+            mce_dbus_get_name_owner_ident(sender));
+
+    if( !dbus_message_get_args(req, &err,
+                               DBUS_TYPE_INT32, &limit,
+                               DBUS_TYPE_INVALID) ) {
+        mce_log(LL_ERR, "Failed to get argument from %s.%s: %s: %s",
+                MCE_REQUEST_IF, MCE_CHARGING_MODE_SET,
+                err.name, err.message);
+        rsp = dbus_message_new_error(req, err.name, err.message);
+        goto EXIT;
+    }
+
+    mce_setting_set_int(MCE_SETTING_CHARGING_LIMIT_ENABLE, limit);
```

Comment:
I think there is a fundamental misunderstanding here. MCE settings are already exposed on D-Bus -> all(?) of these new dbus method calls are unnecessary

For reference

- dbus interface -> https://github.com/sailfishos/mce-dev/blob/master/include/mce/dbus-names.h#L703
- hysteresis control in "mcetool bits" of https://github.com/sailfishos/mce/commit/632daf8b7e4bcdf6c99afd4b38a7a251778e2c21
- use of other settings -> https://github.com/sailfishos/nemo-qml-plugin-systemsettings/blob/master/src/displaysettings.cpp

Reply:
Refactored nemo-dbus side to use the settings API, dropped the offending changes here.